### ISSUE TITLE
Is there an Easy way to measure text?

### ISSUE THREAD

**Issue by Wiladams** [blend2d/blend2d#21]
I am getting lost between font, metrics, glyph buffers and the like.  Is there an obvious way to measure the size of a string?

More specifically, the text bounding box for some text. I know it's an obviously complex thing, but looking for the Cairo equivalent "toy" interface.

**Comment by kobalicek** [comment]
Yeah this is complex and I'm still learning here. But in short, if you want to do anything with text except rendering you have to go through `BLGlyphBuffer`. I have added a small example to [Getting Started](https://blend2d.com/getting-started.html) page that uses glyph-buffer.

Terminology:
  - `BLFontFace` - design font that represents a single `ttf` or `otf` file or a single font in a font collection
  - `BLFontDesignMetrics` - design metrics stored in `BLFontFace`
  - `BLFont` - scaled `BLFontFace` into the user provided `size`
  - `BLFontMetrics` - scaled font metrics that respect user provided `size`
  - `BLGlyphBuffer` - allows to convert unicode text to glyphs
  - `BLGlyphRun` - data structure that contains glyphs and possibly their placement
  - `BLTextMetrics` - metrics of text or glyphs

Basically to measure text you need a glyph-run, to render text you need a glyph-run, basically to do anything with text you need to convert text into glyphs before you can do anything with it, so Blend2D makes this explicit operation and requires users to use `BLGlyphBuffer`.

**Comment by Wiladams** [comment]
The one piece im not sure of is the conversions.  The glyph run gives 'advance, as a big number.  Should i divide that by 64, 92, 96, 120, or some other font related factor to get pixels?

**Comment by kobalicek** [comment]
I think this is a bit messy still. In general GlyphBuffer works with design units, but when you ask for TextMetrics that should get scaled to user units (the size of the font). I have to check how harfbuzz deals with this as I'm not really sure 100% myself how this should work.

**Comment by Wiladams** [comment]
Ah yes, that is the problem.  I had first used text metrics, and was getting design (non-scaled) values.  I'll update and see what's what.  I realize this is a quick and dirty "toy" way to measure text, but it helps at least doing quick and dirty UI stuff.

**Comment by Wiladams** [comment]
I have confirmed that getTextMetrics seems to be giving a scaled result.  The bounding box still does not seem correct though.  The x0 and x1 are correct, but the y0 and y1 return a value of 0.  I use the font height (font.impl.metrics.size) to fillin for that temporarily.

**Comment by kobalicek** [comment]
Yeah this is true, I actually don't know whether text metrics should calculate exact vertical bounding box or whether it should default to something like [-ascent, descent]. I would definitely need some help in this area as I'm no expert on typography (yet:))

**Comment by Wiladams** [comment]
fonts get very tricky very fast with plenty of esoterica mixed in over the centuries.  I'm going to stick with the font.impl.metrics.size thing for now.  Perhaps this is an area left to those fancy font/text layout engines.  Provide all the necessary font, glyph information you parse, and let the true believers build the super correct stuff.  Meanwhile, your 'toy' can do a few things well, and call it a day.

**Comment by Wiladams** [comment]
I'll close this specific issue now and open a separate issue for more specific font measurement challenges.

**Comment by bear24rw** [comment]
>  The x0 and x1 are correct, but the y0 and y1 return a value of 0. I use the font height (font.impl.metrics.size) to fillin for that temporarily.

@Wiladams did this give you a tight bounding box or did you have to scale/adjust it in any way? I'm getting a slightly too tall box using `font.size()` (which uses `font._impl().metrics.size` under the hood)

<img width="640" alt="Image" src="https://github.com/user-attachments/assets/71ac46ed-ebed-4893-aaae-20b364c6cd11" />

**Comment by bear24rw** [comment]
@kobalicek this simple patch gives me a correct looking bbox, I'm only using simple ascii characters without accents or other crazy things so it might not be correct in the general case. What are your thoughts about changing the default behavior to this? It seems less wrong than returning 0.

```patch
diff --git a/src/blend2d/font.cpp b/src/blend2d/font.cpp
index 6d2f2ad..27df533 100644
--- a/src/blend2d/font.cpp
+++ b/src/blend2d/font.cpp
@@ -570,7 +570,7 @@ BL_API_IMPL BLResult blFontGetTextMetrics(const BLFontCore* self, BLGlyphBufferC
 
   out->leadingBearing.reset(lsb, 0);
   out->trailingBearing.reset(rsb, 0);
-  out->boundingBox.reset(glyphBounds[0].x0, 0.0, advance.x - rsb, 0.0);
+  out->boundingBox.reset(glyphBounds[0].x0, glyphBounds[0].y0, advance.x - rsb, glyphBounds[0].y1);
 
   const BLFontMatrix& m = selfI->matrix;
   BLPoint scalePt = BLPoint(m.m00, m.m11);\
```

<img width="615" alt="Image" src="https://github.com/user-attachments/assets/ac6bc9c7-d3e8-406b-8b56-268731f7af78" />

**Comment by bear24rw** [comment]
Ok this seems to give me a tight bbox without modifying the library: `auto height = font.metrics().ascent - font.metrics().descent;`

**Comment by Wiladams** [comment]
Yes, that is pretty much what I do.  I did a bunch of examples of measuring text, and doing alignment and whatnot using .svg and my svandme parser.  Some discussion around here: https://matrix.to/#/!QAErhdTeVDsTzJmIhF:gitter.im/$RmDJJtZI5SqbR1PG6ytj0JdqZH-H04anhYkrWNanq2M?via=gitter.im&via=matrix.org&via=igalia.com
